File: pa2/List.c
//-----------------------------------------------------------------------------
// List.c
// implementation file for the List ADT
//-----------------------------------------------------------------------------


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "List.h"

//Node + new Node Constructor

typedef struct ListNode{
  int data;
  struct ListNode* next;
  struct ListNode* previous;
}ListNode;

//allocates the appropriate amount of memory for new nodes
Node newNode(int data){
  Node N= malloc(sizeof(ListNode));
  N->data=data;
  N->next = NULL;
  N->previous = NULL;
  return N;
}

 /*
  * void freeNode (Node *N)
  * pre: Node to be freed
  * post: frees memory in specified node.
  */
void freeNode (Node *N){
  if (*N!=NULL && N!=NULL){
    free (*N);
    N=NULL;
  }
}


//List + new List Constructor

typedef struct ListObj{
  Node front;
  Node back;
  Node cursor;
  int size;
  int cursorIndex;
}ListObj;


	
List newList(void){
  List L = malloc(sizeof(ListObj));
  L->front= NULL;
  L->back= NULL;
  L->cursor= NULL;
  L->size=0;
  return L;
}
 /*
  * void freeList (List pL)
  * pre: List to be freed
  * post: frees memory of all nodes of specified list,
  * and frees memory of the list itself.
  */
void freeList(List* pL){
   if( pL!=NULL && *pL!=NULL){
    Node N = (*pL)->front; 
    Node temp;
    while(N != NULL){
    temp = N->next;
    N->previous=NULL;
    freeNode(&N);
    N = temp; 
    }
    freeNode(&N);
    free(*pL);
     *pL = NULL;
   }
}

//Access Functions

//returns length of list
int length(List L){
  return L->size;
}


//returns index of cursor
int index(List L){
  if (L->cursor != NULL){
  return L->cursorIndex;
  }
  return -1;
}

/*
 * front(): returns front
 * preconditions : length()>0
 * postconditions: returns data from front
 */

int front(List L){
  return L->front->data;
}


/*
 * back(): returns back
 * preconditions : length()>0
 * postconditions: returns data from back
 */

int back(List L){
  return L->back->data;
}

 /*
  * returns data of cursor
  * precondtions: length()>0, index()>=0
  * postconditions: returns data of cursor
  */
int get(List L){
  if(L->cursor != NULL){
  	return L->cursor->data;
  }
  return -1;
}


/*
 * equals(List)
 * compares two lists to see if they are the same
 * precondtions: another List L
 * postconditions: returns true or false for if the two lists are the same
 */
int equals(List A, List B){
 if (A->size==B->size){
   Node firstFront=A->front;
   Node secondFront=B->front;
   for(moveFront(A); index(A)>=0; moveNext(A)){
     if (firstFront->data!=secondFront->data){
      return 0;
     }
   }
   return 1;
 }
 return 0;
}

// Manipulation procedures

//resets the size + cursor of list.
void clear(List L){
  L->size=0;
  L->cursor=NULL;
}

void moveFront(List L){
  if (L->size > 0){
  	L->cursor=L->front;
  	L->cursorIndex=0;
  }
}

//moves cursor to front
void moveBack(List L){
  if (L->size >0){
	L->cursor=L->back;
	L->cursorIndex=L->size-1;
  }
}

//moves cursor to back
void movePrev(List L){
  if (L->cursor != NULL){
    if (L->cursor==L->front){
    	L->cursor= NULL;
    }
    else{
    	L->cursor= L->cursor->previous;
    	L->cursorIndex--;
    }
  }
}

// moves cursor one step towards the front, if at front, moves cursor
// to undefined,if cursor is undefined, does nothing
void moveNext(List L){
  if (L->cursor != NULL){
    if (L->cursor==L->back){
    	L->cursor= NULL;
    }
    else{
    	L->cursor= L->cursor->next;
    	L->cursorIndex++;
    }
  }
}

/*
 * prepend(int)
 * preconditions: list must be created, paramater int for data to be inserted
 * postconditions: inserts data before front element, if list
 * is empty places data at the front and back (because size=1 after insertion)
 */
void prepend(List L, int data){
  Node N=newNode(data);
  if (L->size==0){
    L->front=N;
    L->back=N;
  }
  else{
    L->front->previous=N;
    N->next=L->front;
    L->front=N;
  }
  if (L->cursor!= NULL){
    L->cursorIndex++;
  }
  L->size++;
}

 /*
  * append(int)
  * preconditions: list must be created, paramater int for data to be inserted
  * postconditions: inserts data after back element, if list
  * is empty places data at front and back (because size=1 after insertion)
  */

void append(List L, int data){
  Node N=newNode(data);
  if (L->size==0){
    L->front=N;
    L->back=N;
  }
  else{
    L->back->next=N;
    N->previous=L->back;
    L->back=N;
  }
  L->size++;
}

/*
 * void insertBefore(int)
 * preconditions: length()>0 index >=0
 * postconditions: inserts new element before cursor.
 */

void insertBefore(List L, int data){
  if (L->cursor!=NULL){
  	if (L->cursor==L->front){
  		prepend(L,data);
  	}
  	else{
      Node N=newNode(data);
  	  L->cursor->previous->next=N;
  	  N->previous=L->cursor->previous;
  	  N->next=L->cursor;
  	  L->cursor->previous=N;
  	  L->size++;
  	  L->cursorIndex++;
  	}
  }
}


/*
 * void insertAfter(int)
 * preconditions: length()>0 index >=0
 * postconditions: inserts new element after cursor.
 */
void insertAfter(List L, int data){
  if (L->cursor!=NULL){
  	if (L->cursor==L->back){
  		append(L,data);
  	}
  	else{
      Node N=newNode(data);
  	  L->cursor->next->previous=N;
  	  N->next=L->cursor->next;
  	  N->previous=L->cursor;
  	  L->cursor->next=N;
  	  L->size++;
  	}
 }
}

/*
 * void deleteFront()
 * preconditions: length()>0
 * postconditions: delete front number 
 */
void deleteFront(List L){
  if (L->cursor==L->front){
    L->cursor=NULL;
  }

  if (L->size>1){
    L->front=L->front->next;
    L->front->previous->next=NULL;
    freeNode(&(L->front->previous));
    L->front->previous=NULL;
    L->size--;
  }
  else if (L->size==1){
    freeNode(&(L->front));
    L->front=NULL;
    L->back=NULL;
    L->size--;
  }

  if (L->cursor!=NULL){
    L->cursorIndex--;
  }
}

/*
 *  void deleteBack() 
 *
 *  preconditions: length()>0
 *  postconditions: delete back number 
 */

void deleteBack(List L){
  if (L->cursor==L->back){
    L->cursor=NULL;
  }

  if (L->size>1){
    L->back=L->back->previous;
    L->back->next->previous=NULL;
    freeNode(&(L->back->next));
    L->back->next=NULL;
    L->size--;
  }
  else if (L->size==1){
    freeNode(&(L->back));
    L->front=NULL;
    L->back=NULL;
    L->size--;
  }
}

/* void delete()
 *  preconditions: length()>0 index()>0
 *  postconditions: delete current item at cursor
 */

void delete(List L){
  if (L->cursor!=NULL){
    if  (L->cursor==L->front){
      deleteFront(L);
    }
    else if (L->cursor == L->back){
      deleteBack(L);
    }
    else{
      L->cursor->previous->next=L->cursor->next;
      L->cursor->next->previous=L->cursor->previous;
      freeNode (&(L->cursor));
      L->cursor=NULL;
      L->size--;
    }
  }
}


//OTHER METHODS

/*
 * copy()
 * precondtions: File to be outputted to, List L
 * postcondtions: prints out list contents to specified file.
 */

void printList(FILE* out, List L){
   for(moveFront(L); index(L)>=0; moveNext(L)){
      fprintf(out,"%d ", get(L));
   }
}

/*
 * copyList()
 * precondtions: List L
 * postcondtions: returns copy of list.
 */

List copyList(List L){
  List copy= newList();
  Node tempN = L->cursor;
  int tempI = L->cursorIndex;
  for(moveFront(L); index(L)>=0; moveNext(L)){
      append(copy,get(L));
   }
   L->cursor = tempN;
   L->cursorIndex = tempI;
   return copy;
}

File: backup/List.java
class List{
    
  private class Node{
  Object entry;
  Node next;
  Node previous;
   
    Node(Object entry){
      this.entry= entry;
  	  next=null;
  	  previous=null;
  	}

    public String toString(){
    	return entry.toString();
    }
  }

//pointers to the front and back elements of list, and a cursor
//int to store size of list and index of cursor.
private Node cursor;
private Node front;
private Node back;
private int size;
private int cursorIndex;

  //Creates empty List
  List (){
  cursor=null;
  front=null;
  back=null;
  size=0;
  }

  //Access Functions

  //returns length of list
  int length(){
    return size;
  }

  //returns index of cursor
  int index(){
    if (cursor != null){
     return cursorIndex;
    }
    return -1;
  }

  // front(): returns front
  // preconditions : length()>0
  // postconditions: returns data from front
 

  Object front(){
  	if (size>0)
  	return  front.entry;
    //case for if front is called on empty list
  	System.out.println("front called on empty list!");
  	return -1;
  }

  // back(): returns back
  // preconditions : length()>0
  // postconditions: returns data from back
  //

  Object back(){
  	if (size>0)
    return back.entry;
    //case for if back is called on empty list
  	System.out.println("back called on empty list!");
  	return -1;
  }
  

  // returns data of cursor
  // precondtions: length()>0, index()>=0
  // postconditions: returns data of cursor

  Object get(){
    if (size!=0 && cursor !=null)
    return cursor.entry;
    //case for if get is called on empty list or cursor is undefined
    return -1;
  }


 // equals(List)
 // compares two lists to see if they are the same
 // precondtions: another List L
 // postconditions: returns true or false for if the two lists are the same


  boolean equals(List L){
  	//if the two lists are of the same size and their front element is the same,
  	//make the comparison.
    if (L.length()==size){
    Node firstFront = front;
    Node secondFront = L.front;
      for (moveFront();index()>=0;moveNext()){
      	if (firstFront.entry!=secondFront.entry){
      		return false;
      	}
      	firstFront=firstFront.next;
      	secondFront=secondFront.next;
      }
      return true;
    }
    return false;
  }

// Manipulation procedures

//resets list
  void clear(){
  	front=null;
  	back=null;
  	cursor=null;
  	size=0;
    cursorIndex=-1;
  }
  
//moves cursor to front
  void moveFront(){
  	if(size>0)
    cursor=front;
    cursorIndex=0;
  }

//moves cursor to back
  void moveBack(){
  	if(size>0)
  	cursor=back;
    cursorIndex=size-1;
  }

// moves cursor one step towards the front, if at front, moves cursor
// to undefined,if cursor is undefined, does nothing
  void movePrev(){
  	if (cursor != null){
  		Node test= cursor;
  		if(test == front){
  			cursor=null;
  		} else {
  	      cursor=test.previous;
  	      cursorIndex--;
  	    }
  	}
  }


// moves cursor one step towards the back, if at back, moves cursor
// to undefined,if cursor is undefined, does nothing
  void moveNext(){
  	if (cursor != null){
  		Node test= cursor;
  		if(test == back){
  			cursor=null;
  		} else {
  	      cursor=test.next;
  	      cursorIndex++;
  	    }
  	}
  }


 // prepend(int)
 // preconditions: list must be created, paramater int for data to be inserted
 // postconditions: inserts data before front element, if list
 // is empty places data at the front and back (because size=1 after insertion)
 
  void prepend(Object entry){
  	//case for list not having any elements
  	if(size==0){
    Node newN = new Node (entry);
    front = newN;
    back= newN;
    size++;
    } else {
    Node n= new Node (entry);
    front.previous=n;
    n.next=front;
    front=n;
    size++;
    cursorIndex++;
    }
  }


 // append(int)
 // preconditions: list must be created, paramater int for data to be inserted
 // postconditions: inserts data after back element, if list
 // is empty places data at front and back (because size=1 after insertion)
 

  void append(Object entry){
  	//case for list not having any elements
  	if(size==0){
    Node newN = new Node (entry);
    front = newN;
    back= newN;
    size++;
    } else {
    Node n= new Node (entry);
    back.next=n;
    n.previous=back;
    back=n;
    size++;
    }
  }


 // void insertBefore(int)
 // preconditions: length()>0 index >=0
 // postconditions: inserts new element before cursor.

  void insertBefore(Object entry){
  	if (cursor != null && size > 0){
      Node n= new Node (entry);
      if (cursor==front){
      	prepend(entry);
      } else {
        cursor.previous.next=n;
        n.previous=cursor.previous;
        n.next=cursor;
        cursor.previous=n;
        size++;
        cursorIndex++;
      }
  	}
  }



 // void insertAfter(int)
 // preconditions: length()>0 index >=0
 // postconditions: inserts new element after cursor.
  void insertAfter(Object entry){
  	if (cursor != null && size > 0){
      Node n= new Node (entry);
      if (cursor==back){
        append(entry);
      } else {
        cursor.next.previous=n;
        n.next=cursor.next;
        n.previous=cursor;
        cursor.next=n;
        size++;
      }
  	}
  }

 // void deleteFront()
 // preconditions: length()>0
 // postconditions: delete front number 
  void deleteFront(){
    if (cursor==front){
      cursor=null;
      cursorIndex = -1;
    }

    if (size >0){
      if (size > 1){
       front=front.next;
       front.previous.next=null;
       front.previous=null;
       size--;
       cursorIndex--;
      } else {
        front=null;
        size--;
      }
      

    }
  }


 //  void deleteBack() 
 //  preconditions: length()>0
 //  postconditions: delete back number
  void deleteBack(){
     if (cursor==back){
      cursor=null;
    } 

    if( size>0){
      if (size > 1){
       back=back.previous;
       back.next.previous=null;
       back.next=null;
      } else {
        back=null;
      }
      size--;
     }	
  }

 // void delete()
 //  preconditions: length()>0 index()>0
 //  postconditions: delete current cursor
  void delete(){
  	if (cursor !=null && size > 0){
      //must include case for if cursor is at front or back
  	  if(cursor==front){
  	  	deleteFront();
  	  } else if (cursor==back){
  	   	deleteBack();
  	  } else {
        cursor.previous.next=cursor.next;
        cursor.next.previous=cursor.previous;
        cursor=null;
        size--;
      }
  	}
  }

  //OTHER METHODS

 // toString() 
 // preconditions: length()>0
 // postconditions: returns string representation of list.
  public String toString(){
  	String toReturn=("");
  	if(size>0){
      for(moveFront();index()>=0;moveNext()){

            toReturn=toReturn+get()+(" ");
       }
    }
    toReturn= toReturn.trim();
     return toReturn;
  }
}